### PR TITLE
Show more information about who did what and when

### LINKED ISSUE #1: Show more information about who did what when.
The log entries created by this extension should show *who* accessed *which* page *when*, and *what* they did (e.g., did they just view the page?  Leave a comment (via the [TeamComments](https://github.com/opentechstrategies/TeamComments) plugin)?  Make an actual edit to the page?

This is to meet the logging needs described in the Torque meetings of [2020-05-14](https://github.com/OpenTechStrategies/torque/wiki/Meeting-Notes#2020-05-14-jeffkarl-meeting-sync-up-and-planning-meeting) (search for "Making logging show *what* the person viewed or did.") and [2020-05-25](https://github.com/OpenTechStrategies/torque/wiki/Meeting-Notes#2020-05-25-12pm-ct-frankkarl-discussion-about-logs-and-devstaging-sites) (search for "Improving the Torque log page.").  Those meeting notes, particularly the second set, also give some links for where to find out more about MediaWiki hooks and how to use them for logging.  @frankduncan and I are available for questions too.
 
*Note:* Secondarily, we may also need to change who has access to these logs.  Right now, they show up on the `Special:Log` page, which any logged-in user can view.  Once the new information listed above is available, it might be that log access needs to be gated, and that would imply greater changes to this plugin than just logging more information.  Let's focus first on the "showing more information" part, though, while we decide whether authorization changes are also called for.

### PR DESCRIPTION
**Pull request by YaxelPerez**
Record user activity and control access to said records.

~~Closes #1~~

Issue #1 
> The log entries created by this extension should show who accessed which page when, and what they did (e.g., did they just view the page? Leave a comment (via the TeamComments plugin)? Make an actual edit to the page?

[2020-05-25 Notes](https://github.com/OpenTechStrategies/torque/wiki/Meeting-Notes#2020-05-25-12pm-ct-frankkarl-discussion-about-logs-and-devstaging-sites)
> Right now we're just setting booleans. We need to update the extension to take anonymous callbacks. Every MediaWiki hook has a set of arguments that get passed to that hook, e.g., "ViewPage" gets a user object and a page object. We would then pass those to the callback, and the callback would return a string that then gets inserted into MediaWiki's database for logged items, which then gets rendered to the Special:Log page.

### CODE REVIEW

**Review by frankduncan — CHANGES_REQUESTED** [review]
Thanks for the change!  Initially looks good.  After the change below, I'll smoke test it quickly and then approve.

Can you update the documentation in the README for this change, as well as adding an example for a common page view hook (we use ArticleViewHeader) to demonstrate how you could use it to log what article was being viewed?

**Comment by kfogel** [comment]
+1 to @frankduncan's comments, but I'd go a bit further:

* The [commit message](https://github.com/OpenTechStrategies/ActivityLog/pull/2/commits/17114f6f67e2f8604cbb6e5ab6af5d2ed97d47f1) itself mentions the Torque meeting notes for a certain date, but doesn't mention issue #1.  This is something I talked about [back on May 7th](https://github.com/OpenTechStrategies/torque/pull/30#issuecomment-625365811), regarding a different PR.  (The *PR* message does mention the issue, but it's the commit message, not the PR message, that goes into the permanent commit history.)

* Also, if you're going to mention specific meeting notes, just give the link to them :-).  Referential convenience is something to keep in mind at all times.

* The commit message goes on to say: "Booleans and strings are also accepted, for instances where just the hook name or a default message would suffice."  This is essentially documentation for the feature -- so instead of being part of the commit message, it should be part of the commit itself, in the form of a documentation update.  This is another general thing to watch out for: putting content in a commit message that actually belongs in the change itself.  The purpose of the commit message is just to memorialize the *change*, and to put a reader in the proper frame of mind to comprehend the corresponding diff.  Information that would be of interest to someone who is looking at the software itself (i.e., not looking at the change) belongs with the software, not tucked away in the commit history.

* As Frank implies, this one commit doesn't "close #1", because there's no instruction to someone deploying this software on how to take advantage of this new functionality.  It needs the actual callback(s) too, along with instructions as to where to put them, so that [Torque](https://github.com/opentechstrategies/torque/) can make use of this feature.

In summary: the code change here may well be correct (I defer to @frankduncan's judgement on that), but it is only a small part of actually resolving the issue.

**Comment by YaxelPerez** [comment]
@frankduncan @kfogel 

Here is what the logs currently look like. I set up ActivityLog to listen for the `ArticleViewHeader` hook:

Screen Shot 2020-06-23 at 3 40 19 PM

```
22:06, 23 June 2020 TestUser1 talk contribs performed an Activity (Visited "Test Page")
```

Instead of just logging "$USER perfomed an Activity (message...)", I'm experimenting right now with a solution described in the [documentation for Special:Log](https://www.mediawiki.org/wiki/Manual:Logging_to_Special:Log#In_i18n/en.json_file) to get the logs to look "native", like the latest log entry generated by [TeamComments](https://github.com/OpenTechStrategies/TeamComments).

MediaWiki already logs edits, and TeamComments (specifically mentioned in issue #1 notes) logs comments without needing ActivityLog. I will add documentation specifying what actions aren't logged already and their corresponding hooks.

**Comment by kfogel** [comment]
Thanks, @YaxelPerez.  This all sounds good.

I wasn't sure exactly what "the latest log entry generated by [TeamComments](https://github.com/OpenTechStrategies/TeamComments)" meant, and I don't know its logging format off the top of my head anyway.  Can you provide an example of what logs that look "native" would look like?

The "so-and-so performed an Activity (activity description in parentheses)" is pretty weird from a user's point of view.  We should aim for super-clear log messages that tell the reader exactly who did what, where, and when in the most direct manner possible.

I don't think we need the "(talk | contribs)" part in each log line.  That's just noise for the reader, or at least for our reader(s).

Also, is there any reason the time-of-day has to come first?

Thinking more generally about the log lines, I'd rather something like this:

    23 Jun 2020 22:06 UTC: [TestUser1](links to TestUser 1) visited [Test Page](links to Test Page).
    23 Jun 2020 22:08 UTC: [TestUser1](links to TestUser 1) added a [comment](link) to [Test Page](link)

etc.  In other words, let's not take the MediaWiki logging format as a given.  We should make this meet the needs of the intended audience as closely as we can.

**Comment by YaxelPerez** [comment]
@kfogel 

> The "so-and-so performed an Activity (activity description in parentheses)" is pretty weird from a user's point of view. We should aim for super-clear log messages that tell the reader exactly who did what, where, and when in the most direct manner possible.

This is exactly what I was trying to say, although I didn't word it as clearly. A "native-looking" log entry would not have the activity description in parentheses.

> Can you provide an example of what logs that look "native" would look like?

```
22:14, 23 June 2020 TestUser1 (talk | contribs) posted a new comment on Test Page (test comment)
```
(this is the TeamComments log entry)

> I don't think we need the "(talk | contribs)" part in each log line. That's just noise for the reader, or at least for our reader(s).

From what I understand, the log message is defined in some i18n JSON file like this:

```
"logentry-foo-bar": "$1 did bar to page $3",
```

`$1` will be substituted with the HTML that corresponds to the user, including the `(talk | contribs` part. Although I agree it is noisy, changing this seems nontrivial at first glance to me.

> Also, is there any reason the time-of-day has to come first?

I believe this is the case for every log entry.

I will take a look at the MediaWiki source to figure out how to change these things.

**Comment by YaxelPerez** [comment]
@kfogel @frankduncan 

callbacks are defined like this now:

```php
// callback is invoked with hook arguments ($article, $outputDone, $pcache in this case)
$wgActivityLogHooksToWatch['ArticleViewHeader'] = function ($article) {
    // callback returns an array with the user (User), action (string), target (Title), and an 
    // optional comment (string)
    return array($article->getContext()->getUser(), 'visited', $article->getTitle());
};
```

and produce a log entry like this:

```
22:51, 1 July 2020 (link to user) (talk|contribs) visited (link to page)
```

**Comment by kfogel** [comment]
Thanks, @YaxelPerez.  I'm going to ask @nolski to review, since he's familiarizing himself with Torque and Torque-related projects right now.

**Comment by kfogel** [comment]
Ah, short delay while @nolski accepts invitation to team.

**Comment by YaxelPerez** [comment]
update: broke torquedata setup on my machine. It had taken a while to set up manually so instead I am spending some time getting ansible working before I pick this back up
edit: still working on this as of jul 9

**Comment by kfogel** [comment]
This PR needs to get reviewed and (assuming no problems) merged.  @frankduncan is now also working on logging, and he happens to be assigned as reviewer already anyway, so Frank: the PR's fate is in your hands :-).

**Comment by Nolski** [comment]
As this is my first time using ActivityLog, I may have other parts of my setup broken. However, I'm noticing that there seems to be some issues with spacing between elements and element `$4` not rendering as shown below.

image

**Comment by frankduncan** [comment]
@Nolski I saw that too last night.  I'm currently working on a more dynamic LogFormatter which should correct it.

**Comment by YaxelPerez** [comment]
@Nolski @frankduncan Not getting this on my machine (fresh mediawiki install). I see an update is WIP, but what did you guys have in your LocalSettings.php?

Screen Shot 2020-07-16 at 6 07 35 PM

**Comment by frankduncan** [comment]
@YaxelPerez It's only if you have legacy ActivityLog messages in the DB from previous deployments.

**Comment by frankduncan** [comment]
I've added a commit that changes the strategy from taking an action to taking a messageKey.  @Nolski @YaxelPerez Please take a look and add a review if you like!

I'm putting this in production today, so if you don't get a chance by the end of the week I'll just merge and we can handle any changes needed in a separate branch.

**Review by frankduncan — APPROVED** [review]
